lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -3472,8 +3472,14 @@ instance IsServerError ErrBalanceTx where
                 [ "The transaction is already balanced. "
                 , "Please send a transaction that requires more inputs/outputs to be picked to be balanced."
                 ]
-        ErrBalanceTxUpdateError (ErrUpdateSealedTxBodyError hint) ->
-            apiError err500 CreatedInvalidTransaction hint
+        ErrBalanceTxUpdateError (ErrExistingKeyWitnesses n) ->
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "The transaction could not be balanced, because it contains "
+                , T.pack (show n), " "
+                , "existing key-witnesses which would be invalid after " -- TODO: I /think/ key-witnesses is an accurate term here, but haven't checked.
```

Comment:
In the ledger it seems the witnesses are all equal:
https://github.com/input-output-hk/cardano-ledger-specs/blob/0eb9b38bb125ec8eedf0f7aabf9e09f9b7bf79f7/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs#L181-L188

But it seems we only get key witnesses from cardano-api:
https://github.com/input-output-hk/cardano-node/blob/0ed04ab8d0a205691219f04623a3d880545a3f34/cardano-api/src/Cardano/Api/Tx.hs#L493-L507

This might actually be a separate problem: any existing non-key witnesses will be lost by the balancing…

Reply:
It turned out that script witnesses, datums, etc were stored in the cardano-api version of the `TxBody`. So all is good.